Add ability to speak next and previous lines/paragraphs when navigating with braille

nvdaes opened the pull request:
<!-- Please read and fill in the following template, for an explanation of the sections see:
https://github.com/nvaccess/nvda/blob/master/projectDocs/dev/githubPullRequestTemplateExplanationAndExamples.md
Please also note that the NVDA project has a Citizen and Contributor Code of Conduct which can be found at https://github.com/nvaccess/nvda/blob/master/CODE_OF_CONDUCT.md. NV Access expects that all contributors and other community members read and abide by the rules set out in this document while participating or contributing to this project. This includes creating or commenting on issues and pull requests.

Please initially open PRs as a draft.
When you would like a review, mark the PR as "ready for review".
See https://github.com/nvaccess/nvda/blob/master/.github/CONTRIBUTING.md.
-->

### Link to issue number:
Fixes #17053 
### Summary of the issue:
Sometimes, it may be desirable tohear lines when navigating with braille.
### Description of user facing changes
- A new command, not assigned to gestures, is available in the braille category to toggle speaking of lines/paragraphs when navigating with braille, to enable or disable this feature. A new checkbox for this has been added in the braille settings panel as well.

### Description of development approach
- A new key for this feature has been added to configSpec.py. The value is a boolean (False by default).
- A new script has been added to globalCommands.py.
- A new function has been added to braille.py, to be used by next and previousLine functions, also modified so that info is expanded to the corresponding read unit, i.e., line or paragraph.
- The braille settings panel has a new checkbox to toggle this feature.
### Testing strategy:
Tested manually in Notepad and Window console, with braille following caret, review cursor and automatic modes.
Also, checked that indentation is reported when reading by lines, not by paragraphs, as expected when unit is specified for speakTextInfo function.
### Known issues with pull request:
None.
### Code Review Checklist:

<!--
This checklist is a reminder of things commonly forgotten in a new PR.
Authors, please do a self-review of this pull-request.
Check items to confirm you have thought about the relevance of the item.
Where items are missing (eg unit / system tests), please explain in the PR.
To check an item `- [ ]` becomes `- [x]`, note spacing.
You can also check the checkboxes after the PR is created.
A detailed explanation of this checklist is available here:
https://github.com/nvaccess/nvda/blob/master/projectDocs/dev/githubPullRequestTemplateExplanationAndExamples.md#code-review-checklist
-->

- [x] Documentation:
  - Change log entry
  - User Documentation
  - Developer / Technical Documentation
  - Context sensitive help for GUI changes
- [x] Testing:
  - Unit tests
  - System (end to end) tests
  - Manual testing
- [x] UX of all users considered:
  - Speech
  - Braille
  - Low Vision
  - Different web browsers
  - Localization in other languages / culture than English
- [x] API is compatible with existing add-ons.
- [x] Security precautions taken.

<!-- Please keep the following -->
@coderabbitai summary

nvdaes committed:
Add ability to speak next and previous lines/paragraphs when navigating with braille

nvdaes committed:
Ensure that braille next and previous line commands won't speak them if braille follows speech output

nvdaes committed:
Use cancelSpeech to improve user experience

nvdaes committed:
Add checkbox to speak lines or paragraphs when navigating with braille

pre-commit-ci[bot] committed:
Pre-commit auto-fix

nvdaes committed:
Update user guide

nvdaes committed:
Update changelog

nvdaes commented:
@burmancomp , feel free to test this again. I've added `cancelSpeech()` before speaking line or paragraph, and a checkbox in the braille settings panel.

nvdaes commented:
I updated the changelog and user guide.
I'll test once more and then I'll request a review from NV Access.

nvdaes committed:
Merge branch 'brailleNextLine' of https://github.com/nvdaes/nvda into brailleNextLine

nvdaes commented:
I feel that changelog.md and userGuide.md are missing.

Thanks @burmancomp . I've pushed changes correctly now and this should be fixed and ready for review.

nvdaes commented:
@burmancomp , happy to see that this is useful for you.

CyrilleB79 commented on user_docs/en/userGuide.md:
```diff
@@ -2297,6 +2297,12 @@ If this is enabled, NVDA will automatically speak the character at the cursor wh
 
+##### Speak character when navigating by line or paragraph {#BrailleSpeakOnNavigating}
+
+If this is enabled, NVDA will automatically speak the current line or paragraph when navigating to it with braille navigation keys.
```
Slight rewording that seems better to me, although that cans probably still be improved.

```suggestion
If this is enabled, NVDA will speak the reached line or paragraph when moving the braille display to the previous or next line.
```

CyrilleB79 commented on source/gui/settingsDialogs.py:
```diff
@@ -4613,6 +4613,14 @@ def makeSettings(self, settingsSizer):
 		self.speakOnRoutingCheckBox.Value = config.conf["braille"]["speakOnRouting"]
 
+		# Translators: The label for a setting in braille settings to speak the current line or paragraph when navigating by them with braille.
+		speakOnNavigatingText = _("Speak &line or paragraph when navigating by them with braille")
```
IMO, the option could be reworded.
* No need to tell "word or paragraph": the feedback depends on the corresponding option
* No need to mention braille since we are already in braille panel IMO

```suggestion
		speakOnNavigatingText = _("Provide speech feedback when moving by &line")
```

CyrilleB79 reviewed (commented):
Some rewording suggestions; this may be discussed.

nvdaes committed:
Remove characters word from section about speaking on navigating in the user guide

nvdaes committed:
Update source/gui/settingsDialogs.py

Co-authored-by: Cyrille Bougot <cyrille.bougot2@laposte.net>

nvdaes committed:
Update user_docs/en/userGuide.md

Co-authored-by: Cyrille Bougot <cyrille.bougot2@laposte.net>

nvdaes commented:
Thanks @CyrilleB79 . I've applied your suggestions.

XLTechie commented on user_docs/en/userGuide.md:
```diff
@@ -2297,6 +2297,12 @@ If this is enabled, NVDA will automatically speak the character at the cursor wh
 
+##### Speak when navigating by line or paragraph {#BrailleSpeakOnNavigating}
+
+If this is enabled, NVDA will speak the reached line or paragraph when moving the braille display to the previous or next line.
```
Using "reached" with this sentence structure, sounds a bit irregular in English. Maybe try:

```suggestion
Enabling this option will cause NVDA to speak lines or paragraphs reached using the braille display's navigation controls.
```

At least, given my understanding of the option as essentially "speech follows braille".

nvdaes committed:
Update user_docs/en/userGuide.md

Co-authored-by: Luke Davis <8139760+XLTechie@users.noreply.github.com>

nvdaes commented:
Thanks @XLTechie . I've applied your suggestion.

LeonarddeR commented on source/braille.py:
```diff
@@ -1659,6 +1659,8 @@ def nextLine(self):
 				return
 		dest.collapse()
 		self._setCursor(dest)
+		dest.expand(self._getReadingUnit())
```
I'm afraid this can become problematic, since you're mutating the TextInfo that was said as cursor before. Instead, I think you should copy the textInfo first.
```suggestion
		copy = dest.copy()
		copy.expand(self._getReadingUnit())
```

LeonarddeR commented on source/braille.py:
```diff
@@ -1677,11 +1679,13 @@ def previousLine(self, start=False):
-					dest.expand(unit)
-			else:  # no page turn support
+			else:
+				# no page turn supportcode ..\nvda
```
It looks like some mess ended up in here
```suggestion
				# no page turn support
```

LeonarddeR commented on source/braille.py:
```diff
@@ -1677,11 +1679,13 @@ def previousLine(self, start=False):
 					pass
 				else:
 					dest = dest.obj.makeTextInfo(textInfos.POSITION_LAST)
-					dest.expand(unit)
```
Why was this removed?

nvdaes replied:
Accidentally. Thanks for pointing it out.

LeonarddeR commented on source/braille.py:
```diff
@@ -1677,11 +1679,13 @@ def previousLine(self, start=False):
 				return
 		dest.collapse()
 		self._setCursor(dest)
+		dest.expand(self._getReadingUnit())
```
Please copy first.

LeonarddeR commented on source/braille.py:
```diff
@@ -3841,3 +3845,18 @@ def _speakOnRouting(info: textInfos.TextInfo):
 	spellTextInfo(info)
+
+
+def _speakOnNavigatingByUnit(info: textInfos.TextInfo):
```
If this function is supposed to speak the reading unit, I think I'd do the expansion in here rather than in the methods that call it. May be even do the copying in here to be on the safe side of things.

LeonarddeR commented on source/braille.py:
```diff
@@ -3841,3 +3845,18 @@ def _speakOnRouting(info: textInfos.TextInfo):
+	# Import late to avoid circular import.
+	from speech.speech import cancelSpeech, speakTextInfo
+
+	readingUnit = handler.buffer.regions[-1]._getReadingUnit()
```
Rather than getting the actual unit this way, I'd add it as a parameter to `_speakOnNavigatingByUnit` instead.

LeonarddeR reviewed (changes requested):
Great to see this being implemented!

nvdaes committed:
Apply suggestions from code review

Co-authored-by: Leonard de Ruijter <3049216+LeonarddeR@users.noreply.github.com>

nvdaes commented:
Thanks @LeonarddeR for your review. I'll mark this as a draft and I'll work on this later.

nvdaes committed:
Apply review suggestions

nvdaes commented:
@LeonarddeR , I think that I've addressed all your comments. I don't have if you can review this again, so I'll mark this as ready for review in case NV Access wants to do it.

nvdaes commented:
@burmancomp , I'll look at this tomorrow.

nvdaes committed:
Appy review suggestions

pre-commit-ci[bot] committed:
Pre-commit auto-fix

SaschaCowley commented on source/braille.py:
```diff
@@ -3841,3 +3844,20 @@ def _speakOnRouting(info: textInfos.TextInfo):
+
+
+def _speakOnNavigatingByUnit(info: textInfos.TextInfo, readingUnit: str) -> None:
+	"""Speaks the reading unit after navigating by it with braille.
+
+	:param info: The TextInfo at the cursor position after navigating.
+	:param readingUnit: The reading unit to expand TextInfo.
+	"""
```
Can the docstring be expanded to note that this won't do anything if the user has disabled this feature?
```suggestion
	"""Speaks the reading unit after navigating by it with braille.
	
	This only has an effect if the user has enabled "Speak when navigating by line or paragraph" in braille settings.

	:param info: The TextInfo at the cursor position after navigating.
	:param readingUnit: The reading unit to expand TextInfo.
	"""
```

SaschaCowley commented on source/braille.py:
```diff
@@ -3841,3 +3844,20 @@ def _speakOnRouting(info: textInfos.TextInfo):
+	# Import late to avoid circular import.
+	from speech.speech import cancelSpeech, speakTextInfo
+
+	cancelSpeech()
+	copy = info.copy()
+	copy.expand(readingUnit)
+	speakTextInfo(copy, unit=readingUnit, reason=controlTypes.OutputReason.CARET)
```
Is there a reason to `cancelSpeech` before copying and expanding the `TextInfo`, rather than immediately before `speakTextInfo`?
```suggestion
	copy = info.copy()
	copy.expand(readingUnit)
	cancelSpeech()
	speakTextInfo(copy, unit=readingUnit, reason=controlTypes.OutputReason.CARET)
```

SaschaCowley reviewed (changes requested):
Thanks @nvdaes, this is looking really good.

nvdaes committed:
Apply suggestions from code review

Co-authored-by: Sascha Cowley <16543535+SaschaCowley@users.noreply.github.com>

pre-commit-ci[bot] committed:
Pre-commit auto-fix

SaschaCowley reviewed (approved):
Thanks for this good work, @nvdaes

nvdaes commented:
Thanks for your work and your review, @SaschaCowley .

Qchristensen commented on user_docs/en/userGuide.md:
```diff
@@ -2297,6 +2297,12 @@ If this is enabled, NVDA will automatically speak the character at the cursor wh
+
+Enabling this option will cause NVDA to speak lines or paragraphs reached using the braille display's navigation controls.
+
+To toggle this option from anywhere, please assign a custom gesture using the [Input Gestures dialog](#InputGestures).
```
While this is not dissimilar to how other commands are written, I know users sometimes have difficulty finding the exact gestures to assign in the input gesture dialog, so I'd like to propose:

```suggestion
To toggle this option from anywhere, please assign a custom gesture to "speakOnNavigatingByUnit" in the "Braille" section of the [Input Gestures dialog](#InputGestures).
```

Qchristensen reviewed (changes requested):
Looks good - one suggestion proposed as a way to improve the user guide moving forward

nvdaes committed:
Update user_docs/en/userGuide.md

Co-authored-by: Quentin Christensen <quentin@nvaccess.org>

nvdaes commented:
Thanks @Qchristensen . I've applied your suggestion.
